geps/gep-1713.md

```diff
@@ -0,0 +1,220 @@
+# GEP-1713: Standard Mechanism to Merge Multiple Gateways
+
+* Issue: [#1713](/kubernetes-sigs/gateway-api/issues/1713)
+* Status: Provisional
+
+(See status definitions [here](overview.md#status).)
+
+## tl;dr
+
+The Gateway Resource is a contention point since it is the only place to attach listeners with certificates. We propose a mechanism to allow distinct Gateway resources to be logically merged.
+
+## Goals
+
+- Define a mechanic to merge multiple Gateways (logically)
+- Define a set of acceptable properties that can be merged and their semantics
+
+## Non-Goals
+
+- Apply a Gateway resource onto N distinct gateways (one to many)
+
+## Introduction
+
+Knative generates on demand per-service certificates using HTTP-01 challenges. There can be O(1000) Knative Services in the cluster which means we have O(1000) distinct certificates. Thus updating a single Gateway resource with this many certificates is a contention point and inhibits horizontal scaling of our controllers.
+
+The spec currently has language to indicate implementations `MAY` merge Gateways resources but the mechanic isn't defined.
+https://github.com/kubernetes-sigs/gateway-api/blob/541e9fc2b3c2f62915cb58dc0ee5e43e4096b3e2/apis/v1beta1/gateway_types.go#L76-L78
+
+## API
+
+We propose adding a new `infrastructure` stanza to the `spec` field of a Gateway. Within `infrastructure` there is an `attachTo` field of type `LocalObjectReference`. Although the use of `LocalObjectReference` allows users to attach to any `kind`, this GEP only defines the behaviour of attaching a Gateway to another Gateway.
+
+
+See [GEP-1867](https://github.com/kubernetes-sigs/gateway-api/pull/1868) for more use cases of `infrastructure`.
+
+#### Go
+
+```go
+type GatewaySpec struct {
+  // Infrastructure defines infrastructure level attributes about this Gateway instance.
+  Infrastructure GatewayInfrastructure `json:"infrastructure"`
+  // ...
+}
+
+type GatewayInfrastructure struct {
+  // AttachTo allows the Gateway to associate itself with another resource.
+  // A common scenario is to reference another Gateway which marks
+  // this Gateway a child of another.
+  AttachTo LocalObjectReference `json:"attachTo"`
+}
+```
+
+#### YAML
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: child-gateway
+  labels:
+    gateway.networking.k8s.io/parent-gateway: parent-gateway
```

Comment:
I don't think this label holds its own weight in the current proposal. Without more context, I would remove it

Reply:
I'll drop this from the GEP. It was really an optimization for implementations to do lookups of related gateways via a list+label selector.

Similar to how the `k8s.io/service-name` label is added to `EndpointSlices`

Reply:
removed